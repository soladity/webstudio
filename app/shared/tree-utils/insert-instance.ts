import produce from "immer";
import { type Instance } from "@webstudio-is/sdk";
import { populateInstance } from "./populate";
import { createInstance } from "./create-instance";

export type InstanceInsertionSpec = {
  component: Instance["component"];
  parentId: Instance["id"];
  id: Instance["id"];
  position: number | "end";
};

export const insertInstanceMutable = (
  rootInstance: Instance,
  instance: Instance,
  spec: Pick<InstanceInsertionSpec, "parentId" | "position">
): boolean => {
  if (spec.parentId !== rootInstance.id) {
    for (const child of rootInstance.children) {
      if (typeof child === "string") continue;
      const hasInserted = insertInstanceMutable(child, instance, spec);
      if (hasInserted === true) return true;
    }
    return false;
  }

  // Inserting inside the selected/hovered instance.
  if (typeof spec.position === "number") {
    rootInstance.children.splice(spec.position, 0, instance);
    return true;
  }

  // Inserting after all children
  if (spec.position === "end") {
    rootInstance.children.push(instance);
    return true;
  }

  return true;
};

export const insertInstance = (
  spec: InstanceInsertionSpec,
  rootInstance: Instance,
  options: { populate?: boolean } = { populate: true }
) => {
  const newInstance = createInstance({
    component: spec.component,
    id: spec.id,
  });
  const populatedInstance = options.populate
    ? populateInstance(newInstance)
    : newInstance;

  const updatedInstance = produce(rootInstance, (draftRootInstance) => {
    insertInstanceMutable(draftRootInstance, populatedInstance, spec);
  });

  return {
    instance: updatedInstance,
    insertedInstance: populatedInstance,
  };
};
