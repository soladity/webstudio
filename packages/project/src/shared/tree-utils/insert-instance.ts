import type { Instance } from "@webstudio-is/project-build";

export type InstanceInsertionSpec = {
  parentId: Instance["id"];
  position: number | "end";
};

export const insertInstanceMutable = (
  rootInstance: Instance,
  instance: Instance,
  spec: undefined | InstanceInsertionSpec
): boolean => {
  if (spec === undefined) {
    return false;
  }
  if (spec.parentId !== rootInstance.id) {
    for (const child of rootInstance.children) {
      if (child.type === "text") {
        continue;
      }
      const hasInserted = insertInstanceMutable(child, instance, spec);
      if (hasInserted === true) {
        return true;
      }
    }
    return false;
  }

  // Inserting inside the selected/hovered instance.
  if (typeof spec.position === "number") {
    rootInstance.children.splice(spec.position, 0, instance);
    return true;
  }

  // Inserting after all children
  if (spec.position === "end") {
    rootInstance.children.push(instance);
    return true;
  }

  return true;
};
