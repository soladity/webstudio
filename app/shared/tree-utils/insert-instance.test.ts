import { type Instance } from "@webstudio-is/sdk";
import { type InstanceInsertionSpec } from ".";
import { insertInstance } from "./insert-instance";

describe("Insert instance", () => {
  test(`inside a nested div using "end": div 1 > div 2 > [div 3]`, () => {
    const instance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [],
        },
      ],
    };
    const instanceInsertionSpec: InstanceInsertionSpec = {
      parentId: "2",
      component: "Box",
      position: "end",
      id: "3",
    };
    const result = insertInstance(instanceInsertionSpec, instance);
    expect(result).toMatchSnapshot();
  });

  test("after a nested div using index: div 1 > div 2, [div 3]", () => {
    const instance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [],
        },
      ],
    };
    const instanceInsertionSpec: InstanceInsertionSpec = {
      parentId: "1",
      component: "Box",
      position: 1,
      id: "3",
    };
    const result = insertInstance(instanceInsertionSpec, instance);
    expect(result).toMatchSnapshot();
  });

  test("before a nested div using index: div 1 > [div 3], div 2", () => {
    const instance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [],
        },
      ],
    };
    const instanceInsertionSpec: InstanceInsertionSpec = {
      parentId: "1",
      component: "Box",
      position: 0,
      id: "3",
    };
    const result = insertInstance(instanceInsertionSpec, instance);
    expect(result).toMatchSnapshot();
  });

  test(`after a nested div using "end": div 1 > div 2, [div 3]`, () => {
    const instance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [],
        },
      ],
    };
    const instanceInsertionSpec: InstanceInsertionSpec = {
      parentId: "1",
      component: "Box",
      position: "end",
      id: "3",
    };
    const result = insertInstance(instanceInsertionSpec, instance);
    expect(result).toMatchSnapshot();
  });
});
