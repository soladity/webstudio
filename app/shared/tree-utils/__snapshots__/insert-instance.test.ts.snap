// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insert instance after a nested div using "end": div 1 > div 2, [div 3] 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
    Object {
      "children": Array [],
      "component": "Box",
      "id": "3",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;

exports[`Insert instance after a nested div using index: div 1 > div 2, [div 3] 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
    Object {
      "children": Array [],
      "component": "Box",
      "id": "3",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;

exports[`Insert instance before a nested div using index: div 1 > [div 3], div 2 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "id": "3",
      "style": Object {},
    },
    Object {
      "children": Array [],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;

exports[`Insert instance inside a nested div using "end": div 1 > div 2 > [div 3] 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "component": "Box",
          "id": "3",
          "style": Object {},
        },
      ],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;
